### ISSUE TITLE
x/tools/gopls: nil Signature (?) deref in Completion

### ISSUE THREAD

**Issue by adonovan** [golang/go#70634]
```
#!stacks
"runtime.sigpanic" && "go/types.(*Signature).Results:=94" && "expectedReturnStmtType"
```
Issue created by [stacks](https://pkg.go.dev/golang.org/x/tools/gopls/internal/telemetry/cmd/stacks).

```go
// expectedReturnStmtType returns the expected type of a return statement.
// Returns nil if enclosingSig is nil.
func expectedReturnStmtType(enclosingSig *types.Signature, node *ast.ReturnStmt, pos token.Pos) types.Type {
	if enclosingSig != nil {
		if resultIdx := exprAtPos(pos, node.Results); resultIdx < len(node.Results) {
			return enclosingSig.Results().At(resultIdx).Type()
		}
	}
	return nil
}
```

This stack `sZBuvA` was [reported by telemetry](https://storage.googleapis.com/prod-telemetry-merged/2024-11-29.json):

- `crash/crash`
- [`runtime.gopanic:+69`](https://cs.opensource.google/go/go/+/go1.23.3:src/runtime/panic.go;l=804)
- `runtime.panicmem:=262`
- [`runtime.sigpanic:+19`](https://cs.opensource.google/go/go/+/go1.23.3:src/runtime/signal_unix.go;l=917)
- [`go/types.(*Signature).Results:=94`](https://cs.opensource.google/go/go/+/go1.23.3:src/go/types/signature.go;l=94)
- [`golang.org/x/tools/gopls/internal/golang/completion.expectedReturnStmtType:+3`](https://cs.opensource.google/go/x/tools/+/gopls/v0.17.0-pre.3:gopls/internal/golang/completion/completion.go;l=2628)
- [`golang.org/x/tools/gopls/internal/golang/completion.expectedCandidate:+37`](https://cs.opensource.google/go/x/tools/+/gopls/v0.17.0-pre.3:gopls/internal/golang/completion/completion.go;l=2316)
- [`golang.org/x/tools/gopls/internal/golang/completion.Completion:+150`](https://cs.opensource.google/go/x/tools/+/gopls/v0.17.0-pre.3:gopls/internal/golang/completion/completion.go;l=651)
- [`golang.org/x/tools/gopls/internal/server.(*server).Completion:+19`](https://cs.opensource.google/go/x/tools/+/gopls/v0.17.0-pre.3:gopls/internal/server/completion.go;l=43)
- [`golang.org/x/tools/gopls/internal/protocol.serverDispatch:+193`](https://cs.opensource.google/go/x/tools/+/gopls/v0.17.0-pre.3:gopls/internal/protocol/tsserver.go;l=363)
- [`golang.org/x/tools/gopls/internal/lsprpc.(*streamServer).ServeStream.ServerHandler.func3:+5`](https://cs.opensource.google/go/x/tools/+/gopls/v0.17.0-pre.3:gopls/internal/protocol/protocol.go;l=160)
- [`golang.org/x/tools/gopls/internal/lsprpc.(*streamServer).ServeStream.handshaker.func4:+52`](https://cs.opensource.google/go/x/tools/+/gopls/v0.17.0-pre.3:gopls/internal/lsprpc/lsprpc.go;l=509)
- [`golang.org/x/tools/gopls/internal/protocol.Handlers.MustReplyHandler.func1:+2`](https://cs.opensource.google/go/x/tools/+/gopls/v0.17.0-pre.3:internal/jsonrpc2/handler.go;l=35)
- [`golang.org/x/tools/gopls/internal/protocol.Handlers.AsyncHandler.func2.2:+3`](https://cs.opensource.google/go/x/tools/+/gopls/v0.17.0-pre.3:internal/jsonrpc2/handler.go;l=104)
- `runtime.goexit:+0`
```
golang.org/x/tools/gopls@v0.17.0-pre.3 go1.23.3 darwin/arm64 other,vscode (10)
```
Dups: mCLZHw

**Comment by adonovan** [comment]
This is a strange one: the nil deref is dominated by a non-nil check.

**Comment by adonovan** [comment]
This stack `mCLZHw` was [reported by telemetry](https://storage.googleapis.com/prod-telemetry-merged/2024-11-29.json):

- `crash/crash`
- [`runtime.gopanic:+69`](https://cs.opensource.google/go/go/+/go1.23.1:src/runtime/panic.go;l=804)
- `runtime.panicmem:=262`
- [`runtime.sigpanic:+9`](https://cs.opensource.google/go/go/+/go1.23.1:src/runtime/signal_windows.go;l=401)
- [`go/types.(*Signature).Results:=94`](https://cs.opensource.google/go/go/+/go1.23.1:src/go/types/signature.go;l=94)
- [`golang.org/x/tools/gopls/internal/golang/completion.expectedReturnStmtType:+3`](https://cs.opensource.google/go/x/tools/+/gopls/v0.17.0-pre.3:gopls/internal/golang/completion/completion.go;l=2628)
- [`golang.org/x/tools/gopls/internal/golang/completion.expectedCandidate:+37`](https://cs.opensource.google/go/x/tools/+/gopls/v0.17.0-pre.3:gopls/internal/golang/completion/completion.go;l=2316)
- [`golang.org/x/tools/gopls/internal/golang/completion.Completion:+150`](https://cs.opensource.google/go/x/tools/+/gopls/v0.17.0-pre.3:gopls/internal/golang/completion/completion.go;l=651)
- [`golang.org/x/tools/gopls/internal/server.(*server).Completion:+19`](https://cs.opensource.google/go/x/tools/+/gopls/v0.17.0-pre.3:gopls/internal/server/completion.go;l=43)
- [`golang.org/x/tools/gopls/internal/protocol.serverDispatch:+193`](https://cs.opensource.google/go/x/tools/+/gopls/v0.17.0-pre.3:gopls/internal/protocol/tsserver.go;l=363)
- [`golang.org/x/tools/gopls/internal/lsprpc.(*streamServer).ServeStream.ServerHandler.func3:+5`](https://cs.opensource.google/go/x/tools/+/gopls/v0.17.0-pre.3:gopls/internal/protocol/protocol.go;l=160)
- [`golang.org/x/tools/gopls/internal/lsprpc.(*streamServer).ServeStream.handshaker.func4:+52`](https://cs.opensource.google/go/x/tools/+/gopls/v0.17.0-pre.3:gopls/internal/lsprpc/lsprpc.go;l=509)
- [`golang.org/x/tools/gopls/internal/protocol.Handlers.MustReplyHandler.func1:+2`](https://cs.opensource.google/go/x/tools/+/gopls/v0.17.0-pre.3:internal/jsonrpc2/handler.go;l=35)
- [`golang.org/x/tools/gopls/internal/protocol.Handlers.AsyncHandler.func2.2:+3`](https://cs.opensource.google/go/x/tools/+/gopls/v0.17.0-pre.3:internal/jsonrpc2/handler.go;l=104)
- `runtime.goexit:+0`
```
golang.org/x/tools/gopls@v0.17.0-pre.3 go1.23.1 windows/amd64 vscode (2)
```

**Comment by adonovan** [comment]
Closely related to https://github.com/golang/go/issues/70636.

**Comment by findleyr** [comment]
There's definitely a bug here though. The condition checked is against len(node.Results), but in the presence of ill typed code, the signature length need not match the length of the return statement.

```
if resultIdx := exprAtPos(pos, node.Results); resultIdx < len(node.Results) {
    return enclosingSig.Results().At(resultIdx).Type()
}
```

My guess is that the compiler is somehow reporting the panic from the wrong call.

**Comment by findleyr** [comment]
Aha, well the bug I spotted is exactly #70636.

Therefore, I'm not sure what _this_ bug is. If it were a bug in the compiler's pclntab, I'd have expected it to _always_ report the wrong function.

**Comment by findleyr** [comment]
For triaging: we expect that this is really a dupe of #70636, somehow obscured by a misreported stack from the compiler. I'll leave this in the v0.17.0 milestone, but if it does not recur in the next prerelease, where #70636 is fixed, I think we can safely transfer this to a compiler/runtime bug.

**Comment by findleyr** [comment]
FWIW, I can reproduce this exact crash with gopls@v0.17.0-pre.3, and the repro from my fix for #70636. Therefore, this is a dupe.

**Comment by adonovan** [comment]
https://go.dev/cl/634038 is a reproducer for the problem that doesn't depend on telemetry.
